src/lib/frontend/frontend.ml

```diff
@@ -398,26 +401,37 @@ module Make(SAT : Sat_solver_sig.S) : S with type sat_env = SAT.t = struct
         {env with sat_env; expl}
       | `Unsat -> env
 
+  let check_step_limit f env =
+    match SAT.get_unknown_reason env.sat_env with
+    | Some (Step_limit _) -> env
+    | _ -> f env
+
   let handle_sat_exn f ?loc env x =
     try f ?loc env x with
     | SAT.Sat t -> {env with res = `Sat t}
     | SAT.Unsat expl -> {env with res = `Unsat; expl = Ex.union expl env.expl}
     | SAT.I_dont_know t -> {env with res = `Unknown t}
   (* The SAT.Timeout exception is not catched. *)
 
+  (* Wraps the function f to check if the step limit is reached (in which case,
+     don't do anything), and then calls the function & catches the
+     exceptions. *)
+  let wrap_f f ?loc env x =
+    check_step_limit (fun env -> handle_sat_exn f ?loc env x) env
+
   let push = handle_sat_exn internal_push
 
   let pop = handle_sat_exn internal_pop
```

Comment:
These also need `wrap_f`

Reply:
I did not put the wrapper on these because I did not want to silently fail (fail = do nothing) on `push` and `pop`.

Reply:
Ah, I did not catch this, but the wrapper should *not* silently fail, it should return an error.

If the underlying state of the SAT solver is invalid, there is no guarantee that `push` and `pop` are correct either, so this could possibly crash as well.

Reply:
Not now `push` can no longer fail with the disabled steps

Reply:
Oh I see your point sorry, we may push a wrong environment; but is it really an issue?

Reply:
The issue is not that `push` can fail. The issue is that if you call another function, say `assume`, and *that* function reaches the step limit, the environment becomes invalid (i.e. arbitrary internal invariants may be broken), and we shouldn't be calling `push` (or `pop`, for that matters) on an invalid environment.

Reply:
> Oh I see your point sorry, we may push a wrong environment; but is it really an issue?

If a step limit is reached, there may or may not be some invariants in the environment that are broken. Without looking at the code, I don't know which invariants are broken, or how calling `push` or `pop` interacts with those broken invariants. It might be that this is perfectly fine, it might not be, but we won't know without tracking all the invariants that can potentially get broken, and even then if we can only call `push` and `pop` it is not very interesting, so we should just prevent that.

Reply:
The sat environment should keep track of its own inconsistency wrt the steps; with this, we could actually save environments, update the step number and make sure we don't ever work with invalid environments. What do you tjink?

Reply:
I am not sure I understand why updating the step number seems to be such a huge issue. Once we have an unknown reason with `Step_limit` updating the step number is not going to change that unknown reason. What am I missing here?

Reply:
Now we use the Limit step as a reason not to keep using the environment, it is not an issue anymore.